src/psij/executors/batch/test/qsub

```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/python3
+import os
+import subprocess
+import sys
+import time
+import traceback
+
+from qlib import *
+from filelock import FileLock
+
+
+job_id = get_next_id()
+
+try:
+    f = open(log_file, 'a')
+    args = [sys.executable, str(my_dir / 'qrun'), job_id, sys.argv[1]]
+    proc = subprocess.Popen(args, stderr=f, stdout=subprocess.DEVNULL, text=True)
+    logger.debug('job %s spawned', job_id)
+    time.sleep(0.05)
+    ec = proc.poll()
+    if ec is not None:
+        msg = 'error launching job; check log for details'
+        logger.warning(msg)
+        sys.stderr.write(msg)
+        sys.exit(1)
+    logger.debug('job %s submitted to runner', job_id)
+    update_job(job_id, 'Q', None)
+    print('Submitted job %s' % job_id, flush=True)
+    logger.debug('job %s qsub exiting', job_id)
+    os._exit(0)
+except Exception as ex:
+    logger.warning('error launching job: %s', ex)
+    traceback.print_exc()
+    os._exit(1)
```

Comment:
Maybe it does the same thing. I remember there being a very specific way to exit that allowed the spawned subprocess to continue and the parent to exit, while others resulted either in the subprocess being killed or the parent waiting for it. It's also possible I'm remembering this incorrectly, but I'd go with not fixing it if it's working.

Reply:
The only things I could find suggested that you actually have to do something special to kill off child processes when the parent exits.

https://stackoverflow.com/questions/19447603/how-to-kill-a-python-child-process-created-with-subprocess-check-output-when-t/19448096#19448096

https://stackoverflow.com/questions/23434842/python-how-to-kill-child-processes-when-parent-dies/23587108

https://unix.stackexchange.com/questions/158727/is-there-any-unix-variant-on-which-a-child-process-dies-with-its-parent

https://stackoverflow.com/questions/5772873/python-spawn-off-a-child-subprocess-detach-and-exit

Reply:
There are subtleties, as your last link points out. The child process can be killed when the parent is killed with a signal, and it's unclear to me exactly what python does as part of its exit cleanup process, although it doesn't look like anything that would affect the child. Maybe. There's the stderr which we want to capture in the parent in case of an error, but not in case of normal qrun execution, so that might need some care.

Also, none of the links mention that if you don't explicitly do `stdout=subprocess.DEVNULL`, the parent process doesn't exit until the child exits. 

That said, it looks like `_exit` is unnecessary.

Reply:
Yeah I guess I would just rather not see `_exit` unless we can confirm it is strictly necessary. 

> Also, none of the links mention that if you don't explicitly do stdout=subprocess.DEVNULL, the parent process doesn't exit until the child exits.

That seems like odd behavior, unless you were piping stdout I guess. What about if you redirect to a file?

Reply:
> > That seems like odd behavior, unless you were piping stdout I guess.
> 
> Nope. You can try it if you want. Remove the `stdout` arg and see what happens.

Even funnier. When you do pipe it, it works fine.

Reply:
I'll give it a try later, but my guess is that the parent really does exit, but since the child is part of the same session, when the parent exits the child is foregrounded. Maybe some extra magic is done to determine whether the child has its output directed to the terminal. I'll see if `Popen(..., start_new_session=True)` does anything.

Reply:
> I'll see if `Popen(..., start_new_session=True)` does anything.

It doesn't. I've tried. And I'm not sure why it should. Unless the python runtime acts as a shell, trying to send signals to processes in the same session when it exits, the session should not be relevant. Additionally, the child process gets no signals except SIGCHLD when the grandchild (the actual job) exits. Even if that was the case, it would not explain why there was a difference between `stdout=DEVNULL` and `stdout=None`.

Reply:
Maybe there's a hint in this...

If you invoke it from a shell, you won't see a difference. It's only visible if you invoke qsub from python with subprocess.run and the out/err piped. In one case, the main program will exit immediately after submitting the job, in the other, it will wait for the grandchild to complete.